test/e2e/virtual/initializingworkspaces/virtualworkspace_test.go

```diff
@@ -0,0 +1,327 @@
+	workspaces, err := sourceKcpTenancyClient.ClusterWorkspaces().List(ctx, metav1.ListOptions{
+		LabelSelector: labels.SelectorFromSet(testLabelSelector).String(),
+	})
+	require.NoError(t, err)
+	workspacesByType := map[string]tenancyv1alpha1.ClusterWorkspace{}
+	for i := range workspaces.Items {
+		workspacesByType[strings.ToLower(workspaces.Items[i].Spec.Type)] = workspaces.Items[i]
+	}
+
+	for initializer, expected := range map[string][]tenancyv1alpha1.ClusterWorkspace{
+		"alpha": {workspacesByType["alpha"], workspacesByType["gamma"]},
+		"beta":  {workspacesByType["beta"], workspacesByType["gamma"]},
+	} {
+		sort.Slice(expected, func(i, j int) bool {
+			return expected[i].UID < expected[j].UID
+		})
+		actual, err := clients[initializer].List(ctx, metav1.ListOptions{}) // no list options, all filtering is implicit
+		require.NoError(t, err)
+		sort.Slice(actual.Items, func(i, j int) bool {
+			return actual.Items[i].UID < actual.Items[j].UID
+		})
+		require.Empty(t, cmp.Diff(expected, actual.Items), "cluster workspace list for initializer %s incorrect", initializer)
+	}
+
+	t.Log("Start WATCH streams to confirm behavior on changes")
+	watchers := map[string]watch.Interface{}
+	for _, initializer := range []string{
+		"alpha", "beta",
+	} {
+		watcher, err := clients[initializer].Watch(ctx, metav1.ListOptions{
+			ResourceVersion: workspaces.ResourceVersion,
+		})
+		require.NoError(t, err)
+		watchers[initializer] = watcher
+	}
+
+	t.Log("Adding a new workspace that both watchers should see")
+	ws, err := sourceKcpTenancyClient.ClusterWorkspaces().Create(ctx, workspaceForType("gamma", testLabelSelector), metav1.CreateOptions{})
+	require.NoError(t, err)
+	require.Eventually(t, func() bool {
+		workspace, err := sourceKcpTenancyClient.ClusterWorkspaces().Get(ctx, ws.Name, metav1.GetOptions{})
+		if err != nil {
+			t.Logf("error listing workspaces: %v", err)
+			return false
+		}
+		return workspacesStuckInInitializing(t, *workspace)
+	}, wait.ForeverTestTimeout, 100*time.Millisecond)
+
+	ws, err = sourceKcpTenancyClient.ClusterWorkspaces().Get(ctx, ws.Name, metav1.GetOptions{})
+	require.NoError(t, err)
+
+	for initializer, watcher := range watchers {
+		select {
+		case evt := <-watcher.ResultChan():
+			require.Equal(t, evt.Type, watch.Added)
+			require.Empty(t, cmp.Diff(evt.Object, ws), "got incorrect object in watch stream for initializer %s", initializer)
+		case <-time.Tick(wait.ForeverTestTimeout):
+			t.Fatalf("never saw a watche event for the %s initializer", initializer)
+		}
+	}
+
+	t.Log("Transitioning the new workspace out of initializing")
+	previous := ws.DeepCopy()
+	oldData, err := json.Marshal(tenancyv1alpha1.ClusterWorkspace{
+		Status: previous.Status,
+	})
+	require.NoError(t, err)
+
+	obj := ws.DeepCopy()
+	obj.Status.Initializers = []tenancyv1alpha1.ClusterWorkspaceInitializer{}
+	obj.Status.Phase = tenancyv1alpha1.ClusterWorkspacePhaseReady
```

Comment:
this is done by the controller, why here manually?

Reply:
You are not testing the whole flow, i.e. reducing test coverage without reason.

Reply:
This isn't a test of that controller, though /shrug - this is snipping a dependency and making the test clearer